### DISCUSSION TITLE
Duplicate Prometheus metric definitions generated from datanodes, which causes failure to publish the metrics to grafana

### DISCUSSION THREAD

**Discussion by sokui** [apache/ozone#6278] [FAQ]
Hi community,

Need some help here. The datanodes in our production generates duplicate prom metrics (duplicate definitions). They seem to be all related to ratis. The following are one of the duplications. This cause the prometheus parsing errors. Do you know what the reason it is, and how to fix it?

\# TYPE ratis_leader_election_lastLeaderElapsedTime gauge
ratis_leader_election_lastLeaderElapsedTime{instance="d09c0aaf-dacd-4ad4-95c7-f616ef2906c6",group="group-72DDBFDDFEB2",} 0.0
\# TYPE ratis_leader_election_lastLeaderElapsedTime gauge
ratis_leader_election_lastLeaderElapsedTime{instance="d09c0aaf-dacd-4ad4-95c7-f616ef2906c6",group="group-82B743C17C01",} 0.0
\# TYPE ratis_leader_election_lastLeaderElapsedTime gauge
ratis_leader_election_lastLeaderElapsedTime{instance="d09c0aaf-dacd-4ad4-95c7-f616ef2906c6",group="group-64B8EF032F77",} 0.0

**Comment by kerneltime** [comment]
@sokui I believe these metrics are generated from the Datanode and each one corresponds to a Raft ring for a pipeline.
The mental model is as follows: Each Datanode uses multiple raft/ratis rings (3 Datanodes each) to replicate data. Thus each Datanode can host multiple leaders and followers in those raft rings. Each raft ring is identified by a group ID `group-72DDBFDDFEB2`. Thus, the same metric can be reported multiple times, once for each group.

**Reply by kerneltime** [reply]
@sokui this should not break Grafana, it is capable of charting the same metric separate by labels. Example: The same metric generated from one node with varying labels
![image](https://github.com/apache/ozone/assets/774455/78d1ecd3-7c9e-4841-9a2a-cc24a1ae3243)

**Comment by sokui** [comment]
> @sokui this should not break Grafana, it is capable of charting the same metric separate by labels. Example: The same metric generated from one node with varying labels ![image](https://private-user-images.githubusercontent.com/774455/308001630-78d1ecd3-7c9e-4841-9a2a-cc24a1ae3243.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MDg5OTk3ODIsIm5iZiI6MTcwODk5OTQ4MiwicGF0aCI6Ii83NzQ0NTUvMzA4MDAxNjMwLTc4ZDFlY2QzLTdjOWUtNDg0MS05YTJhLWNjMjRhMWFlMzI0My5wbmc_WC1BbXotQWxnb3JpdGhtPUFXUzQtSE1BQy1TSEEyNTYmWC1BbXotQ3JlZGVudGlhbD1BS0lBVkNPRFlMU0E1M1BRSzRaQSUyRjIwMjQwMjI3JTJGdXMtZWFzdC0xJTJGczMlMkZhd3M0X3JlcXVlc3QmWC1BbXotRGF0ZT0yMDI0MDIyN1QwMjA0NDJaJlgtQW16LUV4cGlyZXM9MzAwJlgtQW16LVNpZ25hdHVyZT0xNTdhMTQ1ODlhMmRjMGRmOTAzYzhmYTNhMzJkYWIyMjQwZGYyN2U1OGMzYTEwYTQ4N2U5ZjE3NTUxZTNlNzNmJlgtQW16LVNpZ25lZEhlYWRlcnM9aG9zdCZhY3Rvcl9pZD0wJmtleV9pZD0wJnJlcG9faWQ9MCJ9.Q6S9lBtpFqyq6HGbdNFinz5MwCyhJGTiPKvh4vFEprk)

Hi @kerneltime,

Sorry for the confusion. I think the same metrics with different labels are fine. I am talking about the metrics definition, which begins with `# Type`. I think if we define the same metric multiple times, it will trigger this error line in prometheus code https://github.com/prometheus/common/blob/main/expfmt/text_parse.go#L517

Let me update my title to be more clear.

**Comment by adoroszlai** [comment]
@sokui Which version of Ozone is this?  HDDS-5010 fixed similar problem in 1.2.0.  Maybe the fix was incomplete or it got broken since then?

**Reply by sokui** [reply]
It is 1.3.0-SNAPSHOT

**Reply by sokui** [reply]
I see. I think I found the issue now. It is related to ratis prometheus metrics. All my duplicate metrics are from ratis. I think it should be related to these lines: https://github.com/apache/ozone/blob/ozone-1.3.0/hadoop-hdds/framework/src/main/java/org/apache/hadoop/hdds/server/http/PrometheusServlet.java#L65-L67

Do you know how we can fix it?

**Reply by kerneltime** [reply]
@sokui thanks for bringing this up, let me take a look.

**Reply by kerneltime** [reply]
@sokui I used the recent master and spun up a local prometheus setup with the fix from https://github.com/apache/ozone/pull/6288 and I don't see Prometheus complaining about it.

```
# HELP ratis_leader_election_lastLeaderElapsedTime Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-93354247498B.lastLeaderElapsedTime, type=org.apache.ratis.metrics.dropwizard3.Dm3RatisMetricRegistryImpl$$Lambda$545/0x000000084062bc40)
# TYPE ratis_leader_election_lastLeaderElapsedTime gauge
ratis_leader_election_lastLeaderElapsedTime{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.0
# HELP ratis_leader_election_transferLeadershipCount Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-93354247498B.transferLeadershipCount, type=com.codahale.metrics.Counter)
# TYPE ratis_leader_election_transferLeadershipCount gauge
ratis_leader_election_transferLeadershipCount{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.0
# HELP ratis_leader_election_timeoutCount Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-93354247498B.timeoutCount, type=com.codahale.metrics.Counter)
# TYPE ratis_leader_election_timeoutCount gauge
ratis_leader_election_timeoutCount{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 1.0
# HELP ratis_leader_election_electionCount Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-93354247498B.electionCount, type=com.codahale.metrics.Counter)
# TYPE ratis_leader_election_electionCount gauge
ratis_leader_election_electionCount{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 1.0
# HELP ratis_leader_election_lastLeaderElectionElapsedTime Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-93354247498B.lastLeaderElectionElapsedTime, type=org.apache.ratis.metrics.dropwizard3.Dm3RatisMetricRegistryImpl$$Lambda$545/0x000000084062bc40)
# TYPE ratis_leader_election_lastLeaderElectionElapsedTime gauge
ratis_leader_election_lastLeaderElectionElapsedTime{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 1043907.0
# HELP ratis_leader_election_electionTime Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-93354247498B.electionTime, type=com.codahale.metrics.Timer)
# TYPE ratis_leader_election_electionTime summary
ratis_leader_election_electionTime{quantile="0.5",instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.038407917
ratis_leader_election_electionTime{quantile="0.75",instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.038407917
ratis_leader_election_electionTime{quantile="0.95",instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.038407917
ratis_leader_election_electionTime{quantile="0.98",instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.038407917
ratis_leader_election_electionTime{quantile="0.99",instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.038407917
ratis_leader_election_electionTime{quantile="0.999",instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 0.038407917
ratis_leader_election_electionTime_count{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-93354247498B",} 1.0
# HELP ratis_leader_election_lastLeaderElapsedTime Generated from Dropwizard metric import (metric=ratis.leader_election.39855cd0-8120-4791-97cd-a5a0db0b1e0e@group-3F5EC5AA8C10.lastLeaderElapsedTime, type=org.apache.ratis.metrics.dropwizard3.Dm3RatisMetricRegistryImpl$$Lambda$545/0x000000084062bc40)
# TYPE ratis_leader_election_lastLeaderElapsedTime gauge
ratis_leader_election_lastLeaderElapsedTime{instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e",group="group-3F5EC5AA8C10",} 0.0
```

Prometheus log

```
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.616Z caller=main.go:1039 level=info msg="Starting TSDB ..."
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.617Z caller=tls_config.go:274 level=info component=web msg="Listening on" address=[::]:9090
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.617Z caller=tls_config.go:277 level=info component=web msg="TLS is disabled." http2=false address=[::]:9090
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.620Z caller=head.go:606 level=info component=tsdb msg="Replaying on-disk memory mappable chunks if any"
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.620Z caller=head.go:687 level=info component=tsdb msg="On-disk memory mappable chunks replay completed" duration=20.958µs
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.620Z caller=head.go:695 level=info component=tsdb msg="Replaying WAL, this may take a while"
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.673Z caller=head.go:766 level=info component=tsdb msg="WAL segment loaded" segment=0 maxSegment=1
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.674Z caller=head.go:766 level=info component=tsdb msg="WAL segment loaded" segment=1 maxSegment=1
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.674Z caller=head.go:803 level=info component=tsdb msg="WAL replay completed" checkpoint_replay_duration=32.875µs wal_replay_duration=53.948209ms wbl_replay_duration=42ns total_replay_duration=54.027375ms
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.677Z caller=main.go:1060 level=info fs_type=EXT4_SUPER_MAGIC
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.677Z caller=main.go:1063 level=info msg="TSDB started"
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.677Z caller=main.go:1245 level=info msg="Loading configuration file" filename=/etc/prometheus.yml
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.678Z caller=main.go:1282 level=info msg="Completed loading of configuration file" filename=/etc/prometheus.yml totalDuration=864.625µs db_storage=3.083µs remote_storage=1.083µs web_handler=292ns query_engine=1µs scrape=180.042µs scrape_sd=18.458µs notify=417ns notify_sd=875ns rules=1.416µs tracing=3.458µs
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.678Z caller=main.go:1024 level=info msg="Server is ready to receive web requests."
2024-02-27 14:46:13 ts=2024-02-27T22:46:13.678Z caller=manager.go:146 level=info component="rule manager" msg="Starting rule manager..."
```
and the metric shows up in Grafana as expected
```
ratis_leader_election_lastLeaderElapsedTime{component="datanode", exported_instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e", group="group-3F5EC5AA8C10", instance="ozone-datanode-4:9882", job="ozone"}
0

ratis_leader_election_lastLeaderElapsedTime{component="datanode", exported_instance="39855cd0-8120-4791-97cd-a5a0db0b1e0e", group="group-93354247498B", instance="ozone-datanode-4:9882", job="ozone"}
0
```
This seems to be [know issue with dropwizard and prometheus](https://github.com/prometheus/client_java/issues/288) and it would need refactoring of how either Ratis metrics are collected or written out to prometheus endpoint but this does not seem to be causing an issue.
Does prometheus in your setup not report these metrics?

**Reply by sokui** [reply]
If there are no duplicate metric definition, Prometheus won't complain it. Do your metrics contain duplicate definitions and Prometheus does not complain?

Note that, for my issue, it is not always reproducible. My cluster has 40 datanodes, and only 1 datanode does not have any duplicate metric definitions, and all the others have ratis related duplicate metric definitions. This end up only one datanode's metrics shown up in Grafana, and all the other 39 datanodes' metrics are missing.

**Reply by sokui** [reply]
OK. I see your duplicate definition for `# TYPE ratis_leader_election_lastLeaderElapsedTime gauge`. Hmm, interesting. Let me have a further look.

**Reply by sokui** [reply]
So for my cluster, if the datanode has one duplicate metric definition, then all of its metrics cannot be shown at all.

**Reply by kerneltime** [reply]
@sokui can you try out the local setup outlined here https://github.com/apache/ozone/blob/master/hadoop-ozone/dist/src/main/compose/ozone/README.md#add-ons 
Feel free to experiment with how ratis collection is setup. Would be interesting to see why this failing for you and not for local setup or the clusters we monitor.

**Reply by sokui** [reply]
@kerneltime basically, are you saying even with duplicate definitions, your prometheus setup can handle them?

**Reply by kerneltime** [reply]
@sokui yes and if you follow the steps outlined in the [Readme.md](https://github.com/apache/ozone/blob/master/hadoop-ozone/dist/src/main/compose/ozone/README.md#add-ons) you can try it out yourself. That said, we should fix how drop wizard gets setup for multi ratis in Datanode.

**Reply by kerneltime** [reply]
@sokui did you get to look into this a bit more?

**Reply by sokui** [reply]
@kerneltime not yet. We will probably do some test next week.
